### PR TITLE
Feature/concurrent queue seqan

### PR DESCRIPTION
**Pull request by rrahn**
Blocked by ~#947~ and ~#948~

Supersedes #842 

Adds an initial version of the seqan2 concurrent queue. 
~This is still WIP and more tests will come.~
~Also there are some outstanding features that will be added in separate PRs after this is merged.~
~Note for me: make an issue with the remaining tasks.~

### CODE REVIEW

**Comment by h-2** [comment]
Without looking into this in detail: since you adapted SeqAn3 style guide and everything, why put this into contrib and not core/parallel or something like that?

**Comment by rrahn** [comment]
Because there are more adaptions involved. I only adapted things that I had to touch anyway. Also the whole thing needs to be documented and everything, which is something we should have in the future but for now it will fit the purpose as is.

File: include/seqan3/contrib/parallel/concurrent_queue.hpp [hunk 1]
```diff
@@ -0,0 +1,478 @@
...
+        assert(tailPos == headPos + roundSize);
+
+        // get positions of head/tail in current data sequence
+        size_type headIdx = headPos & (roundSize - 1);
+        size_type tailIdx = tailPos & (roundSize - 1);
+
+        // increase capacity
+        data.resize(cap + 1);
+        size_type delta = data.size() - cap;
+        roundSize = static_cast<size_type>(1) << ((data.size() > 1) ? static_cast<size_type>(std::log2(data.size() - 1) + 1) : 1);
```
Comment (marehr) [inline include/seqan3/contrib/parallel/concurrent_queue.hpp]: assert(delta == 1)?

File: test/unit/contrib/parallel/concurrent_queue_test.cpp [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    // Notify queue that no more work is going to be added.
+    queue.close();
+
+    for (auto & t : consumer_pool)
+    {
+        if (t.joinable())
+            t.join();
+    }
+
+    auto v = std::view::iota(0, size_v) | std::view::common;
```
Comment (marehr) [inline test/unit/contrib/parallel/concurrent_queue_test.cpp]: small gauss

File: include/seqan3/contrib/parallel/concurrent_queue.hpp [hunk 3]
```diff
@@ -0,0 +1,478 @@
...
+        return size_impl();
+    }
+    //!\}
+private:
+
+    // Needed in two functions that both acquire exclusive rights.
+    size_type size_impl() const noexcept
+    {
+        size_type mask = roundSize - 1;
+        if ((headPos & mask) <= (tailPos & mask))
```
[the remark this answers is not in the document]
Comment (rrahn) [reply include/seqan3/contrib/parallel/concurrent_queue.hpp]: As discussed. Right now, we just moved it from SeqAn2 into contrib to get it working. Later we should put it into the library with all the documentation.

**Comment by rrahn** [comment]
@marehr in the last commit I renamed the concurrent_queue to buffer_queue in order to be compliant with the standard proposal. I also added functionality for a fixed_size queue. So we have both a growing queue and a queue with fixed capacity.

**Comment by rrahn** [comment]
jenkins was aborted.

File: test/unit/contrib/parallel/buffer_queue_base_test.cpp [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+#include <chrono>
+#include <random>
+#include <string>
+
+#include <seqan3/contrib/parallel/buffer_queue.hpp>
+
+using namespace seqan3;
+
+TEST(concurrent_queue, empty)
```
Comment (marehr) [inline test/unit/contrib/parallel/buffer_queue_base_test.cpp]: name `buffer_queue`

File: test/unit/contrib/parallel/buffer_queue_parallel_test.cpp [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+    // std::chrono::steady_clock::time_point stop = std::chrono::steady_clock::now();
+    // double time_span = std::chrono::duration_cast<std::chrono::duration<double> >(stop - start).count();
+    // std::cout << "throughput: " << static_cast<size_t>(random.size() / time_span) << " values/s" << std::endl;
+
+    EXPECT_EQ(chk_sum, chk_sum2);
+    EXPECT_TRUE(push_status == queue_op_status::success);
+    EXPECT_TRUE(pop_status == queue_op_status::closed);
+}
+
+TEST(buffer_queue, spscx_dynamicsize)
```
Comment (marehr) [inline test/unit/contrib/parallel/buffer_queue_parallel_test.cpp]: xXx

**Pull request merged**